upb/util/compare_test.cc

```diff
@@ -0,0 +1,255 @@
+                                                    int max_depth) {
+  std::string buf1 = ToBinaryPayload(uf1);
+  std::string buf2 = ToBinaryPayload(uf2);
+  return upb_Message_UnknownFieldsAreEqual(buf1.data(), buf1.size(),
+                                           buf2.data(), buf2.size(), max_depth);
+}
+
+upb_UnknownCompareResult CompareUnknown(UnknownFields uf1, UnknownFields uf2) {
+  return CompareUnknownWithMaxDepth(uf1, uf2, 64);
+}
+
+TEST(CompareTest, UnknownFieldsReflexive) {
+  EXPECT_EQ(kUpb_UnknownCompareResult_Equal, CompareUnknown({}, {}));
+  EXPECT_EQ(kUpb_UnknownCompareResult_Equal,
+            CompareUnknown({{1, Varint(123)}, {2, Fixed32(456)}},
+                           {{1, Varint(123)}, {2, Fixed32(456)}}));
+  EXPECT_EQ(
+      kUpb_UnknownCompareResult_Equal,
+      CompareUnknown(
+          {{1, Group({{2, Group({{3, Fixed32(456)}, {4, Fixed64(123)}})}})}},
+          {{1, Group({{2, Group({{3, Fixed32(456)}, {4, Fixed64(123)}})}})}}));
+}
+
+TEST(CompareTest, UnknownFieldsOrdering) {
+  EXPECT_EQ(kUpb_UnknownCompareResult_Equal,
+            CompareUnknown({{1, Varint(111)},
+                            {2, Delimited("ABC")},
+                            {3, Fixed32(456)},
+                            {4, Fixed64(123)},
+                            {5, Group({})}},
+                           {{5, Group({})},
+                            {4, Fixed64(123)},
+                            {3, Fixed32(456)},
+                            {2, Delimited("ABC")},
+                            {1, Varint(111)}}));
+  EXPECT_EQ(kUpb_UnknownCompareResult_NotEqual,
+            CompareUnknown({{1, Varint(111)},
+                            {2, Delimited("ABC")},
+                            {3, Fixed32(456)},
+                            {4, Fixed64(123)},
+                            {5, Group({})}},
+                           {{5, Group({})},
+                            {4, Fixed64(123)},
+                            {3, Fixed32(455)},  // Small difference.
+                            {2, Delimited("ABC")},
+                            {1, Varint(111)}}));
+  EXPECT_EQ(kUpb_UnknownCompareResult_Equal,
+            CompareUnknown({{3, Fixed32(456)}, {4, Fixed64(123)}},
+                           {{4, Fixed64(123)}, {3, Fixed32(456)}}));
+  EXPECT_EQ(
+      kUpb_UnknownCompareResult_Equal,
+      CompareUnknown(
+          {{1, Group({{2, Group({{3, Fixed32(456)}, {4, Fixed64(123)}})}})}},
+          {{1, Group({{2, Group({{4, Fixed64(123)}, {3, Fixed32(456)}})}})}}));
+}
+
+TEST(CompareTest, LongVarint) {
+  EXPECT_EQ(kUpb_UnknownCompareResult_Equal,
+            CompareUnknown({{1, LongVarint(123)}, {2, LongVarint(456)}},
+                           {{1, Varint(123)}, {2, Varint(456)}}));
+  EXPECT_EQ(kUpb_UnknownCompareResult_Equal,
+            CompareUnknown({{2, LongVarint(456)}, {1, LongVarint(123)}},
+                           {{1, Varint(123)}, {2, Varint(456)}}));
+}
+
+TEST(CompareTest, MaxDepth) {
```

Comment:
we should add a fuzzer for this eventually, but not now.

Reply:
Acknowledged.